src/Hash.cc

```diff
@@ -222,16 +212,289 @@ void* HashKey::TakeKey()
 		return CopyKey(key, size);
 	}
 
-void* HashKey::CopyKey(const void* k, int s) const
+char* HashKey::CopyKey(const char* k, size_t s) const
 	{
-	void* k_copy = (void*)new char[s];
+	char* k_copy = new char[s]; // s == 0 is okay, returns non-nil
 	memcpy(k_copy, k, s);
 	return k_copy;
 	}
 
-hash_t HashKey::HashBytes(const void* bytes, int size)
+hash_t HashKey::HashBytes(const void* bytes, size_t size)
 	{
 	return KeyedHash::Hash64(bytes, size);
 	}
 
+void HashKey::Set(bool b)
+	{
+	key_u.b = b;
+	key = reinterpret_cast<char*>(&key_u);
+	size = write_size = sizeof(b);
+	}
+
+void HashKey::Set(int i)
+	{
+	key_u.i = i;
+	key = reinterpret_cast<char*>(&key_u);
+	size = write_size = sizeof(i);
+	}
+
+void HashKey::Set(bro_int_t bi)
+	{
+	key_u.bi = bi;
+	key = reinterpret_cast<char*>(&key_u);
+	size = write_size = sizeof(bi);
+	}
+
+void HashKey::Set(bro_uint_t bu)
+	{
+	key_u.bi = bro_int_t(bu);
+	key = reinterpret_cast<char*>(&key_u);
+	size = write_size = sizeof(bu);
+	}
+
+void HashKey::Set(uint32_t u)
+	{
+	key_u.u32 = u;
+	key = reinterpret_cast<char*>(&key_u);
+	size = write_size = sizeof(u);
+	}
+
+void HashKey::Set(double d)
+	{
+		union {
```

Comment:
What exactly does this union do?

Reply:
I wasn't sure either — I kept it as in the existing code, where it's in the constructor: https://github.com/zeek/zeek/blob/master/src/Hash.cc#L142-L145 ... which has been there for a long time. It seems safe to discard.